File: server/src/routes/jobs.js
const express = require('express');
const { query, queryOne, execute } = require('../db');
const { authRequired, requireRole } = require('../middleware/auth');
const { matchScore } = require('../match');

const router = express.Router();

function jobOut(j, extra = {}) {
  return {
    id: j.id, employer_id: j.employer_id, title: j.title, description: j.description,
    required_skills: JSON.parse(j.required_skills || '[]'), location: j.location,
    job_type: j.job_type, salary_min: j.salary_min, salary_max: j.salary_max,
    status: j.status, created_at: j.created_at, ...extra
  };
}

// Public-ish listing (requires auth, any role)
router.get('/', authRequired, async (req, res) => {
  try {
    const { q, location, skill } = req.query;
    let rows = await query(`SELECT j.*, u.name as employer_name, u.org_name FROM jobs j
      JOIN users u ON u.id = j.employer_id WHERE j.status='open' ORDER BY j.created_at DESC`);

    if (q) {
      const needle = q.toLowerCase();
      rows = rows.filter(r => r.title.toLowerCase().includes(needle) || (r.description || '').toLowerCase().includes(needle));
    }
    if (location) {
      rows = rows.filter(r => (r.location || '').toLowerCase().includes(location.toLowerCase()));
    }
    if (skill) {
      rows = rows.filter(r => JSON.parse(r.required_skills || '[]').some(s => s.toLowerCase().includes(skill.toLowerCase())));
    }

    let candidateSkills = [];
    let candidateExp = 0;
    if (req.user.role === 'candidate') {
      const me = await queryOne('SELECT skills, experience_years FROM users WHERE id = $1', [req.user.id]);
      candidateSkills = JSON.parse(me.skills || '[]');
      candidateExp = me.experience_years || 0;
    }

    const out = rows.map(j => {
      const extra = { employer_name: j.org_name || j.employer_name };
      if (req.user.role === 'candidate') {
        const { score } = matchScore(candidateSkills, JSON.parse(j.required_skills || '[]'), candidateExp);
        extra.match_score = score;
      }
      return jobOut(j, extra);
    });

    if (req.user.role === 'candidate') out.sort((a, b) => (b.match_score || 0) - (a.match_score || 0));
    res.json(out);
  } catch (err) { console.error(err); res.status(500).json({ error: 'Internal server error' }); }
});

router.get('/:id', authRequired, async (req, res) => {
  try {
    const j = await queryOne(`SELECT j.*, u.name as employer_name, u.org_name FROM jobs j
      JOIN users u ON u.id = j.employer_id WHERE j.id = $1`, [req.params.id]);
    if (!j) return res.status(404).json({ error: 'Job not found' });
    res.json(jobOut(j, { employer_name: j.org_name || j.employer_name }));
  } catch (err) { console.error(err); res.status(500).json({ error: 'Internal server error' }); }
});

function stripTags(str) {
  return typeof str === 'string' ? str.replace(/<[^>]*>/g, '').trim() : str;
}

router.post('/', authRequired, requireRole('employer', 'admin', 'placement_agency'), async (req, res) => {
  try {
    const { title, description, required_skills, location, job_type, salary_min, salary_max } = req.body;
    if (!title) return res.status(400).json({ error: 'title is required' });
    const result = await execute(`INSERT INTO jobs (employer_id,title,description,required_skills,location,job_type,salary_min,salary_max)
      VALUES ($1,$2,$3,$4,$5,$6,$7,$8) RETURNING id`, [
      req.user.id, stripTags(title), stripTags(description) || '', JSON.stringify(required_skills || []),
      stripTags(location) || '', job_type || 'Full-time', salary_min || null, salary_max || null
    ]);
    const j = await queryOne('SELECT * FROM jobs WHERE id = $1', [result.rows[0].id]);
    res.status(201).json(jobOut(j));
  } catch (err) { console.error(err); res.status(500).json({ error: 'Internal server error' }); }
});

router.put('/:id', authRequired, requireRole('employer', 'admin', 'placement_agency'), async (req, res) => {
  try {
    const job = await queryOne('SELECT * FROM jobs WHERE id = $1', [req.params.id]);
    if (!job) return res.status(404).json({ error: 'Job not found' });
    if (req.user.role !== 'admin' && job.employer_id !== req.user.id) {
      return res.status(403).json({ error: 'Not your job posting' });
    }
    const { title, description, required_skills, location, job_type, salary_min, salary_max, status } = req.body;
    await execute(`UPDATE jobs SET title=$1, description=$2, required_skills=$3, location=$4, job_type=$5, salary_min=$6, salary_max=$7, status=$8 WHERE id=$9`, [
      stripTags(title ?? job.title), stripTags(description ?? job.description),
      JSON.stringify(required_skills ?? JSON.parse(job.required_skills || '[]')),
      location ?? job.location, job_type ?? job.job_type,
      salary_min ?? job.salary_min, salary_max ?? job.salary_max,
      status ?? job.status, req.params.id
    ]);
    const updated = await queryOne('SELECT * FROM jobs WHERE id = $1', [req.params.id]);
    res.json(jobOut(updated));
  } catch (err) { console.error(err); res.status(500).json({ error: 'Internal server error' }); }
});

router.delete('/:id', authRequired, requireRole('employer', 'admin', 'placement_agency'), async (req, res) => {
  try {
    const job = await queryOne('SELECT * FROM jobs WHERE id = $1', [req.params.id]);
    if (!job) return res.status(404).json({ error: 'Job not found' });
    if (req.user.role !== 'admin' && job.employer_id !== req.user.id) {
      return res.status(403).json({ error: 'Not your job posting' });
    }
    await execute('DELETE FROM jobs WHERE id = $1', [req.params.id]);
    res.status(204).end();
  } catch (err) { console.error(err); res.status(500).json({ error: 'Internal server error' }); }
});

// Employer: list jobs they posted
router.get('/mine/list', authRequired, requireRole('employer', 'admin', 'placement_agency'), async (req, res) => {
  try {
    const rows = await query('SELECT * FROM jobs WHERE employer_id = $1 ORDER BY created_at DESC', [req.user.id]);
    res.json(rows.map(j => jobOut(j)));
  } catch (err) { console.error(err); res.status(500).json({ error: 'Internal server error' }); }
});

module.exports = router;

File: server-python/routers/jobs.py
import json
import re
from fastapi import APIRouter, Depends
from database import query, query_one, execute, execute_returning
from auth import auth_required, require_role

router = APIRouter(prefix="/api/jobs", tags=["jobs"])


def _strip(s):
    return re.sub(r"<[^>]*>", "", s).strip() if isinstance(s, str) else s


def _match_score(candidate_skills, job_skills, exp_years):
    if not job_skills: return 0
    c_lower = {s.lower() for s in candidate_skills}
    j_lower = [s.lower() for s in job_skills]
    matched = sum(1 for s in j_lower if s in c_lower)
    skill_score = matched / len(j_lower) * 70
    exp_score = min(exp_years / 5 * 30, 30) if exp_years else 0
    return round(skill_score + exp_score)


def _job_out(j: dict, extra: dict = None) -> dict:
    rs = j.get("required_skills")
    if isinstance(rs, str):
        try: rs = json.loads(rs)
        except Exception: rs = []
    return {
        "id": j["id"], "employer_id": j["employer_id"], "title": j["title"],
        "description": j.get("description"), "required_skills": rs or [],
        "location": j.get("location"), "job_type": j.get("job_type"),
        "salary_min": j.get("salary_min"), "salary_max": j.get("salary_max"),
        "status": j.get("status"), "created_at": j.get("created_at"),
        **(extra or {}),
    }


@router.get("/mine/list")
async def my_jobs(user: dict = Depends(require_role("employer", "admin", "placement_agency"))):
    rows = await query("SELECT * FROM jobs WHERE employer_id=$1 ORDER BY created_at DESC", [user["id"]])
    return [_job_out(r) for r in rows]


@router.get("")
async def list_jobs(q: str = None, location: str = None, skill: str = None,
                    user: dict = Depends(auth_required)):
    rows = await query("""SELECT j.*, u.name as employer_name, u.org_name FROM jobs j
        JOIN users u ON u.id=j.employer_id WHERE j.status='open' ORDER BY j.created_at DESC""")

    if q:
        needle = q.lower()
        rows = [r for r in rows if needle in (r.get("title") or "").lower() or needle in (r.get("description") or "").lower()]
    if location:
        rows = [r for r in rows if location.lower() in (r.get("location") or "").lower()]
    if skill:
        def has_skill(r):
            rs = r.get("required_skills") or "[]"
            if isinstance(rs, str): rs = json.loads(rs)
            return any(skill.lower() in s.lower() for s in rs)
        rows = [r for r in rows if has_skill(r)]

    candidate_skills, candidate_exp = [], 0
    if user["role"] == "candidate":
        me = await query_one("SELECT skills, experience_years FROM users WHERE id=$1", [user["id"]])
        if me:
            s = me.get("skills")
            candidate_skills = json.loads(s) if isinstance(s, str) else (s or [])
            candidate_exp = me.get("experience_years") or 0

    out = []
    for j in rows:
        extra = {"employer_name": j.get("org_name") or j.get("employer_name")}
        if user["role"] == "candidate":
            rs = j.get("required_skills")
            if isinstance(rs, str): rs = json.loads(rs)
            extra["match_score"] = _match_score(candidate_skills, rs or [], candidate_exp)
        out.append(_job_out(j, extra))

    if user["role"] == "candidate":
        out.sort(key=lambda x: x.get("match_score") or 0, reverse=True)
    return out


@router.get("/{jid}")
async def get_job(jid: int, user: dict = Depends(auth_required)):
    j = await query_one("""SELECT j.*, u.name as employer_name, u.org_name FROM jobs j
        JOIN users u ON u.id=j.employer_id WHERE j.id=$1""", [jid])
    if not j: raise Exception("Not found")
    from fastapi import HTTPException
    if not j: raise HTTPException(404, detail="Job not found")
    return _job_out(j, {"employer_name": j.get("org_name") or j.get("employer_name")})


@router.post("", status_code=201)
async def create_job(body: dict, user: dict = Depends(require_role("employer", "admin", "placement_agency"))):
    from fastapi import HTTPException
    if not body.get("title"): raise HTTPException(400, detail="title is required")
    row = await execute_returning(
        """INSERT INTO jobs (employer_id,title,description,required_skills,location,job_type,salary_min,salary_max)
           VALUES ($1,$2,$3,$4,$5,$6,$7,$8) RETURNING id""",
        [user["id"], _strip(body["title"]), _strip(body.get("description", "")),
         json.dumps(body.get("required_skills") or []),
         _strip(body.get("location", "")), body.get("job_type", "Full-time"),
         body.get("salary_min"), body.get("salary_max")],
    )
    j = await query_one("SELECT * FROM jobs WHERE id=$1", [row["id"]])
    return _job_out(j)


@router.put("/{jid}")
async def update_job(jid: int, body: dict, user: dict = Depends(require_role("employer", "admin", "placement_agency"))):
    from fastapi import HTTPException
    job = await query_one("SELECT * FROM jobs WHERE id=$1", [jid])
    if not job: raise HTTPException(404, detail="Job not found")
    if user["role"] != "admin" and job["employer_id"] != user["id"]:
        raise HTTPException(403, detail="Not your job posting")
    rs = body.get("required_skills")
    if rs is None:
        rs_raw = job.get("required_skills")
        rs = json.loads(rs_raw) if isinstance(rs_raw, str) else (rs_raw or [])
    await execute(
        "UPDATE jobs SET title=$1,description=$2,required_skills=$3,location=$4,job_type=$5,salary_min=$6,salary_max=$7,status=$8 WHERE id=$9",
        [_strip(body.get("title") or job["title"]), _strip(body.get("description") or job.get("description")),
         json.dumps(rs), body.get("location") or job.get("location"),
         body.get("job_type") or job.get("job_type"),
         body.get("salary_min") if "salary_min" in body else job.get("salary_min"),
         body.get("salary_max") if "salary_max" in body else job.get("salary_max"),
         body.get("status") or job.get("status"), jid],
    )
    return _job_out(await query_one("SELECT * FROM jobs WHERE id=$1", [jid]))


@router.delete("/{jid}", status_code=204)
async def delete_job(jid: int, user: dict = Depends(require_role("employer", "admin", "placement_agency"))):
    from fastapi import HTTPException
    job = await query_one("SELECT * FROM jobs WHERE id=$1", [jid])
    if not job: raise HTTPException(404, detail="Job not found")
    if user["role"] != "admin" and job["employer_id"] != user["id"]:
        raise HTTPException(403, detail="Not your job posting")
    await execute("DELETE FROM jobs WHERE id=$1", [jid])
